src/benchmark.rs

```diff
@@ -19,74 +21,72 @@ pub type Context = HashMap<String, Value>;
 pub type Pool = HashMap<String, Client>;
 
-async fn run_iterations(benchmark: Arc<Benchmark>, config: Arc<Config>, concurrency: i64) -> Vec<Report> {
-  let delay = config.rampup / config.concurrency;
-  delay_for(time::Duration::new((delay * concurrency) as u64, 0)).await;
-
-  let mut global_reports = Vec::new();
+async fn run_iteration(benchmark: Arc<Benchmark>, config: Arc<Config>, iteration: i64) -> Vec<Report> {
+  let delay = config.rampup / config.iterations;
+  delay_for(time::Duration::new((delay * iteration) as u64, 0)).await;
 
-  let mut pool: Pool = Pool::new();
+  let mut context: Context = Context::new();
+  let mut reports: Vec<Report> = Vec::new();
 
-  for iteration in 0..config.iterations {
-    let mut context: Context = Context::new();
-    let mut reports: Vec<Report> = Vec::new();
+  let mut pool: Pool = Pool::new(); // TODO: Share pool between all iterations
 
-    context.insert("iteration".to_string(), json!(iteration.to_string()));
-    context.insert("base".to_string(), json!(config.base.to_string()));
-
-    for item in benchmark.iter() {
-      item.execute(&mut context, &mut reports, &mut pool, &config).await;
-    }
+  context.insert("iteration".to_string(), json!(iteration.to_string()));
+  context.insert("base".to_string(), json!(config.base.to_string()));
 
-    global_reports.push(reports);
+  for item in benchmark.iter() {
+    item.execute(&mut context, &mut reports, &mut pool, &config).await;
   }
 
-  global_reports.concat()
+  reports
 }
 
 fn join<S: ToString>(l: Vec<S>, sep: &str) -> String {
-  l.iter().fold("".to_string(),
-                  |a,b| if !a.is_empty() {a+sep} else {a} + &b.to_string()
-                  )
+  l.iter().fold(
+    "".to_string(),
+    |a,b| if !a.is_empty() {a+sep} else {a} + &b.to_string()
+  )
 }
 
 pub fn execute(benchmark_path: &str, report_path_option: Option<&str>, relaxed_interpolations: bool, no_check_certificate: bool, quiet: bool, nanosec: bool) -> Result<Vec<Vec<Report>>, Vec<Vec<Report>>> {
   let config = Arc::new(Config::new(benchmark_path, relaxed_interpolations, no_check_certificate, quiet, nanosec));
 
   if report_path_option.is_some() {
-    println!("{}: {}. Ignoring {} and {} properties...", "Report mode".yellow(), "on".purple(), "threads".yellow(), "iterations".yellow());
+    println!("{}: {}. Ignoring {} and {} properties...", "Report mode".yellow(), "on".purple(), "concurrency".yellow(), "iterations".yellow());
   } else {
-    println!("{} {}", "Threads".yellow(), config.threads.to_string().purple());
+    println!("{} {}", "Concurrency".yellow(), config.concurrency.to_string().purple());
     println!("{} {}", "Iterations".yellow(), config.iterations.to_string().purple());
     println!("{} {}", "Rampup".yellow(), config.rampup.to_string().purple());
   }
 
   println!("{} {}", "Base URL".yellow(), config.base.purple());
   println!();
 
-  let threads = config.threads;
+  let threads = std::cmp::min(num_cpus::get(), config.concurrency as usize);
```

Comment:
this is really the number of optimal threads per machine? most cpus have hyperthreading so they can execute more than one thread per core, very interesting

Reply:
yeah, it looks like it's the number of logical cores, so, it is indeed the optimal number of threads to create